plugins/libplugin.c

```diff
@@ -1266,3 +1309,221 @@ void plugin_main(char *argv[],
 
 	tal_free(plugin);
 }
+
+/*-----------------------------------------------------------------------------
+Sparks
+-----------------------------------------------------------------------------*/
+
+struct plugin_spark_completion {
+	/* Nothing.  Just a tag.  */
+};
+
+struct plugin_spark {
+	struct plugin_spark_completion completion;
+	struct command *cmd;
+	bool completed;
+
+	struct command_result *(*waiter)(struct command *cmd,
+					 void *arg);
+	void *waiter_arg;
+};
+
+/* Adaptor for timer.  */
+struct plugin_spark_timer_adaptor {
+	struct command_result *(*cb)(struct command *cmd,
+				     void *arg);
+	struct command *cmd;
+	void *arg;
+};
+static void
+plugin_spark_timer_adaptor_exec(struct plugin_spark_timer_adaptor *adaptor)
+{
+	struct command_result *res;
+
+	/* Clean up timer.  */
+	(void) timer_complete(adaptor->cmd->plugin);
+
+	tal_steal(tmpctx, adaptor);
+	res = adaptor->cb(adaptor->cmd, adaptor->arg);
+	assert(res == &pending || res == &complete);
+}
+static
+struct command_result *
+plugin_spark_defer(struct command *cmd,
+		   struct command_result *(*cb)(struct command *cmd,
+						void *arg),
+		   void *arg)
+{
+	struct plugin_spark_timer_adaptor *adaptor;
+	struct plugin_timer *timer;
+
+	adaptor = tal(cmd, struct plugin_spark_timer_adaptor);
+	adaptor->cmd = cmd;
+	adaptor->cb = cb;
+	adaptor->arg = arg;
+
+	timer = plugin_timer(cmd->plugin, time_from_sec(0),
+			     &plugin_spark_timer_adaptor_exec,
+			     adaptor);
+	tal_steal(adaptor, timer);
+
+	return command_still_pending(cmd);
+}
+
+/* Core wait function.  */
+struct command_result *
+plugin_wait_spark_(struct command *cmd,
+		   struct plugin_spark **pspark,
+		   struct command_result *(*cb)(struct command *cmd,
+						void *arg),
+		   void *arg)
+{
+	/* Move responsibility for the spark to this function.  */
+	struct plugin_spark *spark = *pspark;
+	*pspark = NULL;
+
```

Comment:
True.  Depends on how you code, I suppose.  I like having a "single-owner" rule, which is what `tal` implicitly encodes, though I am avoiding `tal`-ness in the public interface in case in the future we need to change how sparks are stored (i.e. they may become sub-structures of a larger structure, in which case they are not necessarily `tal`-abble).